cpp/include/rapidsmpf/shuffler/postbox.hpp

```diff
@@ -88,8 +104,16 @@ class PostBox {
   private:
     // TODO: more fine-grained locking e.g. by locking each partition individually.
     mutable std::mutex mutex_;
-    std::unordered_map<PartID, std::unordered_map<ChunkID, Chunk>>
-        pigeonhole_;  ///< Storage for chunks, organized by partition and chunk ID.
```

Comment:
It seems like most of the changes are required because you change the internal storage here.

How about making the `PostBox` templated over the key type instead like this: https://github.com/wence-/rapidsmpf/commit/04e32dc44200b3641f73054112a92fe552f8df11

Then we don't need any changes in the shuffler or these implementations other than to provide a key mapper when constructing the postboxes and inlining the definitions in the header file.

Reply:
I like the idea. I was under the impression that we wanted to change the internal storage as well.

Reply:
I think not in the first round. Not least because where we end up might not want the internal storage change that is used here.

Reply:
Okay. Let me add the changes

Reply:
@wence- I added the suggested templated impl now